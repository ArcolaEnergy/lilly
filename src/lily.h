
class LilyObject {

};

enum LilySpecialKind {
	FALSE,
	TRUE,
	NIL,
};

class LilySpecial : LilyObject {
	int kind;
};

class LilyPair : LilyObject {
	LilyObject* car;
	LilyObject* cdr;
};

class LilyString : LilyObject {
	std::string value;
};

class LilySymbol : LilyObject {
	std::string value;
};

class LilyNumber : LilyObject {
};

class LilyInt64 : LilyNumber {
	int64 value;
};

class LilyDouble : LilyNumber {
	double value;
};

