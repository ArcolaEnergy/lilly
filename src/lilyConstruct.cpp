#include "lilyConstruct.hpp"

LilyObjectPtr WRITELN(LilyObjectPtr v, std::ostream& out) {
	v->onelinePrint(out);
	out << "\n";
	return VOID;
}

LilyObjectPtr WRITELN(LilyObjectPtr v) {
	return WRITELN(v, std::cout);
}


LilyListPtr _LIST(std::initializer_list<LilyObjectPtr> vs) {
	LilyListPtr res= NIL;
	// std::initializer_list does not offer reverse() nor rbegin
	// nor indices in C++11. So use reversion afterwards.
	for (auto v : vs) {
		res= LIST_CONS(v, res);
	}
	return reverse(res);
}

