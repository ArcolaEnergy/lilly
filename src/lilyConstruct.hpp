#include <iostream>
#include <cstdarg>

#define LILY_NEW(classAndArguments) \
	std::shared_ptr<LilyObject>(new classAndArguments)

#define CONS(a,b) LILY_NEW(LilyPair(a,b))
#define NIL LilyNull::singleton()
#define TRUE LilyBoolean::True()
#define FALSE LilyBoolean::False()
#define VOID LilyVoid::singleton()
#define INT(a) LILY_NEW(LilyInt64(a))
#define DOUBLE(a) LILY_NEW(LilyDouble(a))
#define STRING(a) LILY_NEW(LilyString(a))
#define SYMBOL(a) LILY_NEW(LilySymbol(a))

static inline
LilyObjectPtr WRITELN(LilyObjectPtr v, std::ostream& out= std::cout) {
	v->onelinePrint(out);
	out << "\n";
	return VOID;
}

static inline
LilyObjectPtr _LIST(std::initializer_list<LilyObjectPtr> vs) {
	LilyObjectPtr res= NIL;
	// std::initializer_list does not offer reverse() nor rbegin
	// nor indices in C++11. So use reversion afterwards.
	for (auto v : vs) {
		res= CONS(v, res);
	}
	return reverse(res);
}

#define LIST(...) _LIST({ __VA_ARGS__ })


