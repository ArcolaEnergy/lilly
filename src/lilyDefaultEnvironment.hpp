#include "lily.hpp"


LilyListPtr lilyDefaultEnvironment();

