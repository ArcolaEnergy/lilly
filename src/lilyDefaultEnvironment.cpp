#include "lily.hpp"
#include "lilyConstruct.hpp"
#include "lilyDefaultEnvironment.hpp"


template <typename LilyT, typename T>
static
T _lilyFold(LilyList* vs, std::function<T(LilyT*,T)> fn, T start) {
	T res= start;
	while (true) {
		if (auto pair= dynamic_cast<LilyPair*>(vs)) {
			if (auto v=  UNWRAP_AS(LilyT, pair->_car)) {
				res= fn(v, res);
				vs= LIST_UNWRAP(pair->_cdr);
			} else {
				throw std::logic_error("not an integer");
			}
		} else if (dynamic_cast<LilyNull*>(vs)) {
			return res;
		} else {
			throw std::logic_error(STR("not a proper list, ending in: "
						   << show(vs)));
		}
	}
}

#define DEF_FOLD_UP_NATIVE(name, LilyT, T, OP, START)			\
	static								\
	LilyObjectPtr name(LilyListPtr* vs,				\
			   LilyListPtr* _ctx,				\
			   LilyListPtr* _cont) {			\
		return LILY_NEW						\
			(LilyT(_lilyFold<LilyT, T>			\
			       (XLIST_UNWRAP(*vs),			\
				[](LilyT* num, T res) -> T {		\
				       /* XX check for overflow! */	\
				       return res OP num->value;	\
			        },					\
				START)));				\
	}

DEF_FOLD_UP_NATIVE(lilyAdd, LilyInt64, int64_t, +, 0);
DEF_FOLD_UP_NATIVE(lilyMult, LilyInt64, int64_t, *, 1);

// LONESTART is used when there's only one argument
#define DEF_FOLD_DOWN_NATIVE(name, LilyT, T, OP, LONESTART)		\
	static								\
	LilyObjectPtr name(LilyListPtr* vs,				\
			   LilyListPtr* _ctx,				\
			   LilyListPtr* _cont) {			\
	auto fn= [](LilyT* num, T res) -> T {				\
		/* XX check for overflow? */				\
		return res OP num->value;				\
	};								\
	LilyList* _vs= &**vs;						\
	if (is_LilyNull(_vs))						\
		throw std::logic_error(#OP ": wrong number of arguments"); \
	T res;								\
	LilyList* r= &*(_vs->rest());					\
	if (is_LilyNull(r))						\
		res= fn(XUNWRAP_AS(LilyT, _vs->first()), LONESTART);	\
	else								\
		res= _lilyFold<LilyT, T>				\
			(r, fn, XUNWRAP_AS(LilyT, _vs->first())->value); \
	return LILY_NEW(LilyT(res));					\
	}

DEF_FOLD_DOWN_NATIVE(lilySub, LilyInt64, int64_t, -, 0);
DEF_FOLD_DOWN_NATIVE(lilyQuotient, LilyInt64, int64_t, /, 1);
DEF_FOLD_DOWN_NATIVE(lilyRemainder, LilyInt64, int64_t, %, 1);

// DEF_NATIVE(lilyDiv, LilyInt64, int64_t, /); // generic


static
LilyObjectPtr lilyCons(LilyListPtr* vs,
		       LilyListPtr* _ctx,
		       LilyListPtr* _cont) {
	// WARN("cons: "<<show(vs));
	LETU_AS(vs0, LilyPair, *vs);
	if (vs0) {
		LETU_AS(vs1, LilyPair, vs0->_cdr);
		if (vs1) {
			LETU_AS(vs2, LilyNull, vs1->_cdr);
			if (vs2) {
				return CONS(vs0->_car, vs1->_car);
			}
		}
	}
	throw std::logic_error("cons needs 2 arguments");
}

static LilyObjectPtr lilyCar(LilyListPtr* vs,
			     LilyListPtr* _ctx,
			     LilyListPtr* _cont) {
	return apply1ary("car", [](LilyObjectPtr v) {
			LETU_AS(p, LilyPair, v);
			if (p)
				return p->_car;
			else
				throw std::logic_error(STR("not a pair: "
							   << show(v)));
		}, vs);
}

static LilyObjectPtr lilyCdr(LilyListPtr* vs,
			     LilyListPtr* _ctx,
			     LilyListPtr* _cont) {
	return apply1ary("cdr", [](LilyObjectPtr v) {
			LETU_AS(p, LilyPair, v);
			if (p)
				return p->_cdr;
			else
				throw std::logic_error(STR("not a pair: "
							   << show(v)));
		}, vs);
}

static LilyObjectPtr lilyQuote(LilyListPtr* es,
			       LilyListPtr* _ctx,
			       LilyListPtr* _cont) {
	return apply1ary("quote", [](LilyObjectPtr e) {
			DEBUGWARN("quote: " << show(e));
			return e;
		}, es);
}

static LilyObjectPtr lilyLength(LilyListPtr* arguments,
				LilyListPtr* _ctx,
				LilyListPtr* _cont) {
	// arguments == ((12 13 14))
	LET_AS(_arguments, LilyList, *arguments);
	if (is_LilyNull(&*(_arguments->rest()))) {
		// we have just one argument, cool.
		auto l= _arguments->first(); // (12 13 14) or "hello" or something
		int64_t len= 0;
		while (true) {
			LET_AS(p, LilyPair, l);
			if (p) {
				len++; // check overflow? 64bit int is pretty large though :)
				l= p->cdr();
			} else {
				LET_AS(null, LilyNull, l);
				if (null) {
					break;
				} else {
					throw std::logic_error
						(STR("not a list: "
						     << show(_arguments->first())));
				}
			}
		}
		return INT(len);
	} else {
		throw std::logic_error("length received more than one argument");
	}
		
}

static LilyObjectPtr lilyList(LilyListPtr* arguments,
			      LilyListPtr* _ctx,
			      LilyListPtr* _cont) {
	return *arguments;
}

static LilyObjectPtr lilyReverse(LilyListPtr* arguments,
				 LilyListPtr* _ctx,
				 LilyListPtr* _cont) {
	return apply1ary("reverse", [](LilyObjectPtr v) {
			return reverse(v);
		}, arguments);
}


static LilyObjectPtr lilyDefine(LilyListPtr* es,
				LilyListPtr* ctx,
				LilyListPtr* cont) {
	LET_AS(es0, LilyPair, *es);
	if (!es0)
		throw std::logic_error("define needs at least 1 argument");
	// match 1st argument
	auto var_or_pair= es0->car();
	LET_AS(var, LilySymbol, var_or_pair);
	if (var) {
		LET_AS(es1, LilyPair, es0->cdr());
		if (!es1)
			throw std::logic_error("define: if getting a symbol as 1st argument, need one more argument");
		LET_AS(end, LilyNull, es1->cdr());
		if (!end)
			throw std::logic_error("define: if getting a symbol as 1st argument, accepting just one more argument");
		auto expr= es1->car();
		// รง
	}
	LET_AS(bindform, LilyPair, var_or_pair);
	if (bindform) {
		LET_AS(var, LilySymbol, bindform->first());
		if (var) {
			throw std::logic_error("XX functions not yet implemented");
		} else {
			throw std::logic_error(STR("define: expecting symbol, got: "
						   << show(bindform->first())));
		}
	}

	throw std::logic_error("define needs a symbol as the first argument");
	const LilyObjectPtr& _es1= es0->cdr();
	LET_AS(es1, LilyPair, _es1);
	if (es1) {
		//const LilyObjectPtr& 
	} else {
		// set variable to void (or remove it altogether?)
		//XXX
	}
}


static LilyObjectPtr lilyBegin(LilyListPtr* es,
			       LilyListPtr* ctx,
			       LilyListPtr* cont) {
	// evaluate all es in turn; drop all results before the last
	return NIL;
}

#define ENTRY(str, proc) PAIR(SYMBOL(str, false), NATIVE_PROCEDURE(proc, str))

LilyListPtr lilyDefaultEnvironment() {
	static LilyListPtr env= LIST(
		ENTRY("+", lilyAdd),
		ENTRY("*", lilyMult),
		ENTRY("-", lilySub),
		ENTRY("quotient", lilyQuotient),
		ENTRY("remainder", lilyRemainder),
		// ENTRY("integer./", lilyDiv),
		ENTRY("cons", lilyCons),
		ENTRY("car", lilyCar),
		ENTRY("cdr", lilyCdr),
		ENTRY("quote", lilyQuote),
		ENTRY("list", lilyList),
		ENTRY("length", lilyLength),
		ENTRY("reverse", lilyReverse),
		ENTRY("define", lilyDefine),
		ENTRY("begin", lilyBegin),
		);
	return env;
}


