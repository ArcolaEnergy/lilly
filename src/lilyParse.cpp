#include <functional>
#include <assert.h>
#include "lily.hpp"
#include "lilyConstruct.hpp"
#include "lilyParse.hpp"
#include "lilyUtil.hpp"

enum class ParseResultCode : char {
	// the ones hard-coded in parse.cpp
	Success=0,
	UnexpectedEof=1,
	UnexpectedString=2,

	Whitespace, // parser found whitespace at point, not object
	MissingInput,
	Unimplemented,

	UnknownSyntax,
	UnknownBangSpecial,
	UnknownSpecial,
	NotAnInteger,
	NotANumber,
	Int64Overflow,
	DivisionByZero,
	NotASymbol,
	InvalidDottedList,
};

const char* ParseResultCode_string (ParseResultCode c) {
	switch (c) {
	case ParseResultCode::Success: return "Success";
	case ParseResultCode::UnexpectedEof: return "UnexpectedEof";
	case ParseResultCode::UnexpectedString: return "UnexpectedString";

	case ParseResultCode::Whitespace: return "Whitespace"; // shouldn't happen, meta
	case ParseResultCode::MissingInput: return "MissingInput";
	case ParseResultCode::Unimplemented: return "Unimplemented";

	case ParseResultCode::UnknownSyntax: return "UnknownSyntax";
	case ParseResultCode::UnknownBangSpecial: return "UnknownBangSpecial";
	case ParseResultCode::UnknownSpecial: return "UnknownSpecial";
	case ParseResultCode::NotAnInteger: return "NotAnInteger";
	case ParseResultCode::NotANumber: return "NotANumber";
	case ParseResultCode::Int64Overflow: return "Int64Overflow";
	case ParseResultCode::DivisionByZero: return "DivisionByZero";
	case ParseResultCode::NotASymbol: return "NotASymbol";
	case ParseResultCode::InvalidDottedList: return "InvalidDottedList";
	}
	// how can I make the compiler warn about missing cases but
	// otherwise be silent, without the throw ?
	throw std::logic_error("missing case");
}


bool needsSymbolQuoting (char c) {
	return !(isWordChar(c)
		 || (c == '!')
		 || (c == '?')
		 || (c == '.')
		 || (c == ':')
		 || (c == '/')
		 || (c == '%')
		 || (c == '$')
		 || (c == '-')
		 || (c == '+')
		 || (c == '*')
		 || (c == '_')
		 || (c == '&')
		 || (c == '=')
		 || (c == '<')
		 || (c == '>')
		 || (c == '#')
		);
}

// all the character that will introduce another token after a number,
// or also symbol or keyword?
bool isSeparation (char c) {
	return isWhitespace(c)
		|| needsSymbolQuoting(c);
}

static
S skipWhitespaceAndComments (S s) {
	s= skipWhitespaceOnly(s);
	if (s.isNull())
		return s;
	char c= s.first();
	if (c == ';') {
		return skipWhitespaceAndComments(skipUntilAfterEol(s));
	} else {
		auto rest= s.rest();
		if ((c == '#') && (! rest.isNull()) && (rest.first() == '|'))
			return skipWhitespaceAndComments
				(skipUntilAfter(rest.rest(), "|#"));
		else
			return s;
	}
}


typedef ParseResult<LilyObjectPtr> PR;

static
PR parseError(S s, ParseResultCode error) {
	return PR(VOID, s.setError(error));
}



// s is after '#'
static
PR parseHashitem(S s) {
	if (s.isNull())
		return parseError(s, ParseResultCode::UnexpectedEof);
	char c1= s.first();
	auto r= s.rest();
	if (c1 == '!') {
		// special object
		r= expectString(r, "void");
		if (r.success()) {
			if (isWordEndBoundary(r))
				return PR(VOID, r);
		}
		return parseError(r, ParseResultCode::UnknownBangSpecial);
	} else if (c1 == '|') {
		// skipWhitespaceAndComments should have eliminated
		// this case
		throw std::logic_error("bug");
	} else {
		if (isWordEndBoundary(r)) {
			if (c1 == 'f') {
				return PR(FALSE, r);
			} else if (c1 == 't') {
				return PR(TRUE, r);
			} else {
				return parseError(s, ParseResultCode::UnknownSpecial);
			}
		} else {
			return parseError(r, ParseResultCode::UnknownSpecial);
		}
	}
}

// s is after '"' or '|'
static
PR parseStringLike(S s,
		   char quoteChar,
		   std::function<PR(const std::string&, S rest)>
		   cont) {
	std::string str;
	while (true) {
		if (s.isNull())
			return parseError(s, ParseResultCode::UnexpectedEof);
		char c= s.first();
		s= s.rest();
		if (c == quoteChar)
			return cont(str, s);
		if (c == '\\') {
			if (s.isNull())
				return parseError(s, ParseResultCode::UnexpectedEof);
			c = s.first();
			s= s.rest();
			if (c=='n') {
				c='\n';
			} else if (c=='r') {
				c='\r';
			} else if (c=='t') {
				c='\t';
			} else if (c=='f') {
				c='\f';
			} else if (c=='0') {
				c='\0';
			}
			// XX handle lots of other cases?

			// else leave c as is
		}
		// XX handle unicode?
		str.push_back(c);
	}
}



PR parsePositiveInteger(S s) {
	if (s.isNull())
		return parseError(s, ParseResultCode::MissingInput);
	int64_t res=0;
	bool isnumber= false;
	bool isoverflow= false;
	while (true) {
		if (s.isNull())
			break;
		char c= s.first();
		if (isDigit(c)) {
			s=s.rest();
			auto d = c - '0';
			try {
				res= lily_add(lily_mul(res, 10), d);
			} catch (std::overflow_error) {
				isoverflow= true;
			}
			isnumber=true;
		} else {
			// equivalent to isWordEndBoundary, isNull
			// check already done
			if (isWordChar(c))
				isnumber=false;
			break;
		}
	}
	if (isnumber) {
		if (isoverflow)
			return parseError(s, ParseResultCode::Int64Overflow);
		else
			return PR(INT(res), s);
	} else {
		return parseError(s, ParseResultCode::NotAnInteger);
	}
}

// could have a leading + or -; does not verify boundary after the
// integer
PR parseInteger(S s) {
	if (s.isNull())
		return parseError(s, ParseResultCode::MissingInput);
	char c0= s.first();
	bool isneg= false;
	if (c0=='-') {
		isneg= true;
	} else if (c0=='+') {
		isneg= false;
	} else {
		return parsePositiveInteger(s);
	}
	auto pr= parsePositiveInteger(s.rest());
	if (!pr.success())
		return pr;
	if (isneg) {
		LETU_AS(v, LilyInt64, pr.value());
		try {
			return PR(INT(lily_negate(v->value)), pr.remainder());
		} catch (std::overflow_error) {
			return parseError(s, ParseResultCode::Int64Overflow);
		}
	} else {
		return pr;
	}
}

PR parseNumber(S s) {
	auto result= parseInteger(s);
	if (! result.success())
		return result;

	if (result.remainder().isNull())
		return result;
	char c= result.remainder().first();
	DEBUGWARN("     c="<<c);
	switch (c) {
	case '/': {
		// 1/-3 should be parsed as a symbol, at least
		// Gambit does that, hence we have to do this
		auto s= result.remainder().rest();
		if (s.isNull())
			goto notanumber;
		auto num2= parsePositiveInteger(s);
		if (num2.success()) {
			// Oh, and we have to check again what
			// follows it. Evil syntax? Tokenizer
			// actually makes sense for *this*
			// reason. Should have proper boundary
			// detection functions to achieve the
			// same (todo).
			s= num2.remainder();
			if ((! s.isNull()) && (s.first() == '/'))
				goto notanumber;
			XLETU_AS(n, LilyInt64, result.value());
			XLETU_AS(d, LilyInt64, num2.value());
			try {
				// todo location keeping
				result= PR(Divide(n, d), s);
				goto successsofar;
			} catch (LilyDivisionByZeroError) {
				// XX ever report start, not end?
				goto notanumber;
			}
		} else {
			// returning num would be wrong now
			// that we know it would be valid as a
			// fractional up to this point? I
			// mean, we can and should fall back
			// to parsing as a symbol now. Weird
			// case?
			return num2;
		}
	}
	default:
		goto successsofar;
	}
successsofar:
	// now there must be a proper separation after
	// the number, otherwise it's not one.
	if ((result.remainder().isNull())
	    || isSeparation(result.remainder().first()))
		return result;
notanumber:
	return parseError(s, ParseResultCode::NotANumber);
}

// unquoted (no '|' around it) symbols
PR parseSymbol(S s) {
	std::string str;
	while (true) {
		if (s.isNull())
			break;
		char c= s.first();
		if (!needsSymbolQuoting(c))
			str.push_back(c);
		else
			break;
		s=s.rest();
	}
	auto len= str.length();
	if ((len == 1) && str[0] == '.')
		return parseError(s, ParseResultCode::NotASymbol);
	else if ((len > 1) && (str[len-1] == ':')) {
		str.pop_back();
		return PR(KEYWORD(str, false), s);
	}
	// check (str[0] == ':') for CL style keywords, give them another type?
	else
		return PR(SYMBOL(str, false), s);
}

// move to header file? but then also need to move the above classes.
PR lilyParse (S s);

// s is after the '('
PR parseList(S s) {
	s= skipWhitespaceAndComments(s);
	if (s.isNull())
		return parseError(s, ParseResultCode::UnexpectedEof);
	char c=s.first();
	if (c==')')
		return PR(NIL, s.rest());
	if (c=='.') {
		auto s1=s.rest();
		if (s1.isNull())
			return parseError(s1, ParseResultCode::UnexpectedEof);
		auto s2= skipWhitespaceAndComments(s1);
		if (s2.position() > s1.position()) {
			// dot stands by itself, i.e. dotted pair; expect 1 element then ")"
			auto r1= lilyParse(s2);
			if (!r1.success())
				return r1; // cutting away all the stored stuff. OK?
			auto s2= r1.remainder();
			s2= skipWhitespaceAndComments(s2);
			if (s2.isNull())
				return parseError(s2, ParseResultCode::UnexpectedEof);
			char c2= s2.first();
			if (c2 == ')')
				return PR(r1.value(), s2.rest());
			else
				return parseError(s2, ParseResultCode::InvalidDottedList);
		} else {
			// must be a symbol starting with a dot, right?  XX
			auto res= parseSymbol(s);
			if (!res.success()) {
				if (res.error() == ParseResultCode::NotASymbol)
					return parseError(s2, ParseResultCode::InvalidDottedList);
				else
					return res; // ditto cutting away
			}
			auto tail= parseList(res.remainder());
			if (!tail.success())
				return tail; // ditto, and, this is our 'failure monad'
			return PR(CONS(res.value(), tail.value()),
				  tail.remainder());
		}
	} else {
		// parse an item, then the remainder of the list
		PR res= lilyParse(s);
		if (!res.success())
			return res;
		auto tail= parseList(res.remainder());
		if (!tail.success())
			return tail;
		return PR(CONS(res.value(), tail.value()),
			  tail.remainder());
	}
}


static
PR newString(const std::string& str, S rest) {
	return PR(STRING(str), rest);
}
static
PR newSymbolOrKeyword(const std::string& str, S rest) {
	if ((! rest.isNull()) && (rest.first() == ':'))
		return PR(KEYWORD(str, true), rest.rest());
	else
		return PR(SYMBOL(str, true), rest);
}


PR lilyParse (S s) {
	s= skipWhitespaceAndComments (s);
	if (s.isNull())
		return parseError(s, ParseResultCode::MissingInput);
	char c= s.first();
	auto s1= s.rest();
	LilyObjectPtr* special_symbol;
	if (c=='(') {
		return parseList(s1);
	} else if (c=='#') {
		return parseHashitem(s1);
	} else if (c=='"') {
		return parseStringLike(s1, '"', newString);
	} else if (c=='|') {
		return parseStringLike(s1, '|', newSymbolOrKeyword);
	} else if (c==';') {
		// until the end of the line; if s is 1 line then that
		// will be eof, but make it generic so actually check:
		return lilyParse(skipUntilAfterEol(s1));
	} else if (c=='\'') {
		special_symbol= &lilySymbol_quote; goto special;
	} else if (c=='`') {
		special_symbol= &lilySymbol_quasiquote; goto special;
	} else if (c==',') {
		special_symbol= &lilySymbol_unquote; goto special;
	} else {
		// attempt numbers, plain symbol (correct in that order?)
		auto v= parseNumber(s);
		if (v.success())
			return v;
		// if it was a proper number but just overflowed,
		// return it as such:
		if (v.error() == ParseResultCode::Int64Overflow)
			return v;
		// not parseable as a number (even potentially
		// bignum), try symbol:
		v= parseSymbol(s);
		if (v.success())
			return v;
		// return parseError(s, ParseResultCode::UnknownSyntax);

		// actually return the symbol parsing error since
		// symbol is what it remains to be interpreted as
		// anyway?
		return v;
	}
	throw std::logic_error("unreachable");
special:
	auto r= lilyParse(s1);
	return PR(CONS(*special_symbol, CONS(r.value(), NIL)),
		  r.remainder());
}

// convenience function
LilyObjectPtr lilyParse (std::string s, bool requireTotal) /* noexcept */ {
	PR r= lilyParse(StringCursor(&s));
	if (r.success()) {
		if (requireTotal) {
		    // check that there's nothing after the parsed
		    // expression
		    auto end= skipWhitespaceAndComments(r.remainder());
		    if (end.isNull())
			    return r.value();
		    else
			    return PARSEERROR("non-whitespace after expression",
					      end.position());
		} else {
			return r.value();
		}
	} else {
		// exception? error result? Or s-expression for
		// producing the error? :
		ParseResultCode code= r.error();
		return PARSEERROR(ParseResultCode_string(code),
				  r.remainder().position());
	}
}
