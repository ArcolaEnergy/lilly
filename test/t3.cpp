#include <lily.hpp>
#include <lilyConstruct.hpp>
#include <lilyParse.hpp>
#include <lilyDefaultEnvironment.hpp>

auto environment= lilyDefaultEnvironment();

void note(const char* s) {
	std::cerr << "---- " << s << " ----\n"; //
	std::cout << "---- " << s << " ----\n";
}

void e(const char* codestring) {
	auto codeobject= lilyParse(std::string(codestring));
	try {
		auto result= eval(codeobject, environment);
		std::cout << result->typeName() << ": "; WRITELN(result);
	} catch (std::logic_error& e) {
		std::cout << "ERR: " << e.what() << std::endl;
	}
}

void _e(const char* codestring) {
	auto codeobject= lilyParse(std::string(codestring));
	auto result= eval(codeobject, environment);
	std::cout << result->typeName() << ": "; WRITELN(result);
}

int main () {
	note("string and symbol pointer comparison");
	// Yes, == on the pointer is a pointer comparison, and
	// interning works
	{
		const char* p= "Hello";
		auto s1= STRING(p);
		auto s2= STRING(p);
		auto s3= STRING(std::string("Hel") + std::string("lo"));
		WRITELN(LIST(s1,s1,BOOLEAN(s1==s1)));
		WRITELN(LIST(s1,s2,BOOLEAN(s1==s2)));
		WRITELN(LIST(s1,s3,BOOLEAN(s1==s3)));
		// WRITELN(LIST(s1,s3,BOOLEAN(*s1==*s3)));
		// no match for ‘operator==’ (operand types are ‘LilyObject’ and ‘LilyObject’)
		//  of course.
	}
	{
		auto s1= SYMBOL("Hello");
		auto s2= SYMBOL("Hello");
		auto s3= SYMBOL(std::string("Hel") + std::string("lo"));
		auto s4= SYMBOL("Hello2");
		WRITELN(LIST(s1,s1,BOOLEAN(s1==s1)));
		WRITELN(LIST(s1,s2,BOOLEAN(s1==s2)));
		WRITELN(LIST(s1,s3,BOOLEAN(s1==s3)));
		// WRITELN(LIST(s1,s3,BOOLEAN(*s1==*s3)));
		WRITELN(LIST(s1,s4,BOOLEAN(s1==s4)));
	}

	note("literals");
	e("1233");
	e(" \"1233\" ");
	e("#t");
	e("#!void");

	note("variable reference");
	e(" + "); // NativeProcedure
	e(" q "); // undefined variable

	note("calls");
	e("(+)"); // 0
	e("(+ 19)"); // 19
	e("(+ 10 20)"); // 30
	e("(- 10 20)"); // -10
	e("(- 10 20 -5)"); // -5
	e("(+ 10 3 4)"); // 17
	e("(+ (+ 10 3) 4)"); // 17
	e("(+ (* 10 3) 4)"); // 34
	_e("(quote x)"); // x
	e("'y"); // y
	e("(quote (quote x))"); // 'x
	e("(cons 12 (cons 13 '()))");
	e("(list 12 13 14)");
	e("(cons (+ ((car (cons * +)) (+ 10 1) (- 3 2)) 4) 'hey)"); // (15 . hey)
	e("(cons (car 13) ())"); // tests argument eval order, too, not standardized
	e("(car 4 5)");
	e("(car)");
	e("(cdr (cons 3 6))");
	e("(cdr)");
	e("(quotient 10 3)"); // 3
	e("(quotient -10 3)"); // -3
	e("(remainder -10 3)"); // -1
	e("(modulo -10 3)"); // 2
	e("(/ 10 3)"); // 3.33333 ?
	e("(integer./ 10 3)"); // 3
	e("(double./ 10 3)"); // error
	e("(double./ 10. 3.)"); // 3.3333
	e("(cons 10 (list 11 12 '()))"); // (10 11 12 ())

	note("invalid");
	e("()"); // empty call
	e("(+ 10 . 20)"); // improper list -- XX say 'argument list'
	e("(+ . 10)"); // improper list -- XX say 'argument list'
	e("(10 20)"); // not a function error
}
