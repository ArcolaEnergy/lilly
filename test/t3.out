---- string and symbol pointer comparison ----
("Hello" "Hello" #t)
("Hello" "Hello" #f)
("Hello" "Hello" #f)
(Hello Hello #t)
(Hello Hello #t)
(Hello Hello #t)
(Hello Hello2 #f)
---- invalid evaluation rules ----
> (12 . 45 67)
ERR: ill-formed expression: #<parse-error "InvalidDottedList" 9>
---- literals ----
> 1233
Int64: 1233
> "1233"
String: "1233"
> #t
Boolean: #t
> #!void
Void: #!void
---- variable reference ----
> +
NativeProcedure: #<native-procedure +>
> q
ERR: variable not bound: q
---- calls ----
> (+)
Int64: 0
> (+ 19)
Int64: 19
> (+ 10 20)
Int64: 30
> (- 10 20)
Int64: -10
> (- 10 20 -5)
Int64: -5
> (+ 10 3 4)
Int64: 17
> (+ (+ 10 3) 4)
Int64: 17
> (+ (* 10 3) 4)
Int64: 34
> 'y
Symbol: x
> 'y
Symbol: y
> ''x
Pair: 'x
> (cons 12 (cons 13 '()))
Pair: (12 13)
> (length (cons 12 (cons 13 '())))
Int64: 2
> (length '())
Int64: 0
> (length '(1 89 3))
Int64: 3
> (length 1 89 3)
ERR: length received more than one argument
---- list ----
> (list)
Null: ()
> (list 14)
Pair: (14)
> (list 12 13 14)
Pair: (12 13 14)
> (length "hello")
ERR: not a list: "hello"
> (length (list 12 13 14))
Int64: 3
> (reverse (list 12 13 14))
Pair: (14 13 12)
> (cons (+ ((car (cons * +)) (+ 10 1) (- 3 2)) 4) 'hey)
Pair: (15 . hey)
> (cons (car 13) ())
ERR: not a pair: 13
> (car 4 5)
ERR: car needs 1 argument
> (car)
ERR: car needs 1 argument
> (cdr (cons 3 6))
Int64: 6
> (cdr)
ERR: cdr needs 1 argument
> (cons 10 (list 11 12 '()))
Pair: (10 11 12 ())
---- arithmetic ----
> (-10)
ERR: not a function: -10
> (- 10)
Int64: -10
> (- 10 . 33)
ERR: improper list: 33
> (-)
ERR: -: wrong number of arguments
> (quotient 10 3)
Int64: 3
> (quotient -10 3)
Int64: -3
> (quotient 30)
Int64: 0
> (remainder 30)
Int64: 1
> (remainder -10 3)
Int64: -1
> (modulo -10 3)
Int64: 2
> (/ 10 3)
Fractional64: 10/3
> (integer./ 10 3)
Fractional64: 10/3
> (double./ 10 3)
ERR: not a LilyDouble: 10
> (double./ 10. 3.)
Double: 3.33333333333333
---- arithmetic r5rs tests ----
> (modulo 13 4)
Int64: 1
> (remainder 13 4)
Int64: 1
> (modulo -13 4)
Int64: 3
> (remainder -13 4)
Int64: -1
> (modulo 13 -4)
Int64: -3
> (remainder 13 -4)
Int64: 1
> (modulo -13 -4)
Int64: -1
> (remainder -13 -4)
Int64: -1
---- number tower ----
> (+ 5 2/3)
Fractional64: 17/3
> (+ 100 1/2)
Fractional64: 201/2
> (+ 1/2 100)
Fractional64: 201/2
> (+ 5/4 2/3)
Fractional64: 23/12
> (+ 1/6 1/3)
Fractional64: 1/2
> (+ 5/6 1/3)
Fractional64: 7/6
> (+ 1/4 0.5)
Double: 0.75
> (+ 0.5 1/4)
Double: 0.75
> (- 1/4 0.5)
Double: -0.25
> (- 0.5 1/4)
Double: 0.25
> (- 100)
Int64: -100
> (+ 100)
Int64: 100
> (- 3/4)
Fractional64: -3/4
> (+ 3/4)
Fractional64: 3/4
> (- 100.)
Double: -100.
> (+ 100.)
Double: 100.
> (* 100)
Int64: 100
> (/ 100)
Fractional64: 1/100
> (* 3/4)
Fractional64: 3/4
> (/ 3/4)
Fractional64: 4/3
> (* 100.)
Double: 100.
> (/ 100.)
Double: 0.01
> (+ 2.25 0.5)
Double: 2.75
> (- 2.25 0.5)
Double: 1.75
> (* 2.25 0.5)
Double: 1.125
> (/ 2.25 0.5)
Double: 4.5
> (* 1/4 2/3)
Fractional64: 1/6
> (* 1/4 2/3 5)
Fractional64: 5/6
> (* 1/4 2/3 5 2.)
Double: 1.66666666666667
> (/ 1/4 2/3)
Fractional64: 3/8
> (/ 3/4 2)
Fractional64: 3/8
> (/ 2 3/4)
Fractional64: 8/3
---- exact->inexact ----
> (exact->inexact 3)
Double: 3.
> (exact->inexact 0.3)
Double: 0.3
> (exact->inexact 3/2)
Double: 1.5
---- invalid ----
> ()
ERR: empty call
> (+ 10 . 20)
ERR: improper list: 20
> (+ . 10)
ERR: improper list: 10
> (10 20)
ERR: not a function: 10
---- chars ----
> (string->list "Hello\0 World\b\a\f\t\r\n")
(#\H #\e #\l #\l #\o #\nul #\space #\W #\o #\r #\l #\d #\backspace #\alarm #\page #\tab #\return #\newline)
> (list #\f #\esc #\return)
(#\f #\esc #\return)
> (list->string (list #\f #\esc #\return #\xff #\ufffe #\U0001effe))
"f\33\r\377\ufffe\U0001effe"
> (string->list "f\33\r\377\ufffe\U0001effe")
(#\f #\esc #\return #\xff #\ufffe #\U0001effe)
